discord-scripts/channel-management.ts

```diff
@@ -0,0 +1,185 @@
+    }
+
+    // move channel to archived-channel category and send out transcript to interaction
+    discordClient.on("interactionCreate", async (interaction) => {
+      if (
+        !interaction.isCommand() ||
+        interaction.commandName !== "archive-channel"
+      ) {
+        return
+      }
+      if (!interaction.guild || !(interaction.channel instanceof TextChannel)) {
+        return
+      }
+
+      try {
+        const allMessages = await fetchAllMessages(interaction.channel)
+        robot.logger.info(`Total messages fetched: ${allMessages}`)
+
+        const messagesContent = allMessages
+          .reverse()
+          .map(
+            (m) =>
+              `${m.createdAt.toLocaleString()}: ${m.author.username}: ${
+                m.content
+              }`,
+          )
+          .join("\n")
+
+        const filePath = `${interaction.channel.name}_transcript.txt`
+        await writeFile(filePath, messagesContent, "utf-8")
+
+        // check for or create archived category
+        let archivedCategory = interaction.guild.channels.cache.find(
+          (c) =>
+            c.type === ChannelType.GuildCategory &&
+            c.name.toLowerCase() === "archived-channels",
+        )
+        if (!archivedCategory) {
+          archivedCategory = await interaction.guild.channels.create({
+            name: "archived-channels",
+            type: ChannelType.GuildCategory,
+          })
+        }
+
+        // move channel and set permissions
+        if (archivedCategory) {
+          await interaction.channel.setParent(archivedCategory.id)
+          await interaction.channel.permissionOverwrites.edit(
+            interaction.guild.id,
+            { SendMessages: false },
+          )
+          await interaction.reply(
+            "**Channel archived, locked and moved to #archived-channel category**",
+          )
+
+          // upload chat transcript to channel and then delete it
+          const fileAttachment = new AttachmentBuilder(filePath)
+          await interaction.channel
+            .send({
+              content: "**Here is a transcript of the channel messages:**",
+              files: [fileAttachment],
+            })
+            .then(() => unlink(filePath))
+            .catch((error) =>
+              robot.logger.error(`Failed to delete file: ${error}`),
+            )
+
+          robot.logger.info(
+            "Channel archived and locked successfully, messages saved.",
+          )
+        }
+      } catch (error) {
+        robot.logger.error(`An error occurred: ${error}`)
+      }
+    })
+
+    // move channel back to defense category on Unarchived
+    discordClient.on("interactionCreate", async (interaction) => {
+      if (
+        !interaction.isCommand() ||
+        interaction.commandName !== "unarchive-channel"
+      ) {
+        return
+      }
+      if (!interaction.guild) {
+        return
+      }
+      try {
+        if (interaction.channel instanceof TextChannel) {
```

Comment:
We also need to limit this interaction to only be executable in the Archive: Defense category or someone *will* accidentally trigger it and create chaos :grimacing: